### ISSUE TITLE
Firebase Messaging & Rest Api Call Plugin

### ISSUE THREAD

**Issue by AnoxySoftware** [thingsboard/thingsboard#682]
Hi,

I want to basically use Firebase Messaging to send Push Notifications to certain mobile phones.
I've set up Firebase Console and some topics.

Now I only need to do the FCM callback to send the Push, which I can't do.
So basically I want to send the topic value from the device as telemetry data and when this is received the callback to Firebase should be made

I've configured a new Plugin called firebase as Rest Api Call Plugin with these settings:
```
url scheme: https
server:fcm.googleapis.com
port: 443
base path: /fcm/send
No Auth
Request header: Authentikation:key=<FCM Server KEY>
Request header: Content-Type=application/json
```

Now I also set up a rule:

> Filter: Message Type Filter
> Plugin: Firebase (above plugin)
> Action: Send Message 
> Body Template: 
> {"notification":{"title": "Title","text": "Your Text","sound": "default","badge": "1","color": "#990000"},"priority" : "high","to":$topic.valueAsString}

I'm pushing new Telemetry data to a device with the following:

`mosquitto_pub -d -h "server" -p 1883 -t "v1/devices/me/telemetry" -u "$AccessToken" -m '{"topic":"myTopic"}'`

I'm seeing in the Dashboard the new Telemetry data, but I can't get the push on my device.
When sending the push from Postman it works fine.
So questions:

1. Did I do everything ok ?
2. How can I debug if the Rule is actually triggering and if it is sending the plugin action
3. How can I debug what the Rest Api Plugin is sending (Headers/Body/URL etc)

**Comment by AnoxySoftware** [comment]
OK I did debug a bit more.
Issue is that the Body Template is not Json Encoded.
Can we actually add in the Rest API Plugin an option for Encoding Body as JSON and not escaping Double Quotes ?

**Comment by faringvoz** [comment]
you could solve it?

**Comment by AnoxySoftware** [comment]
I've created a PHP script actually that takes the Post BODY and formats it in JSON and does the request from there.
Not the best solution, but it works

**Comment by faringvoz** [comment]
you could show me an example of how to send the data, I have the same problem.
Body Template:
  {
    apikey: "7dfcb3f64d6c4dc042fc9dceba2ec032",
    from: "13056767097",
    to: "14082900459",
    msj: "prueba truway 17"

  }

**Comment by AnoxySoftware** [comment]
I've actually created a Plugin called Firebase.
In there I'm setting the base path to the PHP script I've created.
I'm setting 2 request headers:

1. FcmKey --> My Firebase Key
2. Content-Type --> text/html; charset=utf-8

Then I have a Firebase Rule
Which uses this body: (basically whenever a new attribute is uploaded for key "msg" it will send the push. I'm uploading the attributes "msg" and "topic" together to create the message and I have a Client Shared Attribute called "deviceName" to use the name of the device in the push as the title)

$msg.valueAsString
$topic.valueAsString
default
$cs.deviceName

so parameters are message, topic, sound and title for Push

Attached the script as a txt format as I can't use php attachments here
[fcm.txt](https://github.com/thingsboard/thingsboard/files/1920641/fcm.txt)
